CHANGELOG.md

```diff
@@ -3,6 +3,12 @@ Changelog
 
 See also the [GitHub releases page](https://github.com/FriendsOfSymfony/FOSHttpCache/releases).
 
+2.15.4
+------
+
+* Add flag `fallbackToSmaxage` to `CustomTtlListener` to allow controlling fallback to s-maxage if custom TTL header is not defined on the response.
+* Fix: Do not call store if Response object is not longer cacheable after event listeners.
```

Comment:
how severe was this bug? did we actually end up storing uncacheable things? did we also deliver them?

that is: did this bug lead to serving data that should not have been cached? or is it "just" a performance fix?

Reply:
> how severe was this bug? did we actually end up storing uncacheable things? did we also deliver them?

It didn't have any effects or bugs to the end user or the cached items as there was no cachelifetime for this items and so they were stored. I did not see that any files where created.

It only did trigger a unnecessary write operations in the PSR6Store, as there is a counter for a autoPrune behaviour here: https://github.com/Toflar/psr6-symfony-http-cache-store/blob/9b3eb5b5f449a886b3ce202ddf3705e3d2ac79c9/src/Psr6Store.php#L459 which is normally only called when public responses are cached. Now it was in our case also for "private" endpoints called and so autpruning happened a lot of more times and this unnecessary write counter increases.

Reply:
thanks for the clarification. what if a listener changes other cache-control directives like adding a no-store?

i suggest we are a bit more detailed to not unnecessairy scare people looking at the changelog:

```suggestion
* Performance fix: Do not call store if Response object is not longer cacheable after event listeners. Calling the store method does not actually store the response if the cache lifetime is 0, but wastes resources as the behaviour is not expected.
```

Reply:
@dbu That is a good point. Calling `$response->setPrivate();` or `$response->headers->addCacheControlDirective('no-store')` inside a PreHandle Listener behaviour is now changed:

Before: `$response->setPrivate()` did not have any effects and the response was still cached if it was cacheable before.

After: `$response->setPrivate();` will make it uncacheable also for the http cache.

Same for no-store part. The were only validate before the listeners but never after the listeners were called: https://github.com/symfony/symfony/blob/04f9e6dc07cc228fb3023579c0a63a6ae06a4674/src/Symfony/Component/HttpFoundation/Response.php#L544.

So making a cached response uncachable didn't work before. But I did not find any builtin listener which is modifying that even the UserContextListener does not, they handle all via `Vary` header correctly and so keep the cache-control public as they should.

Reply:
I would expect if I set inside a preHandle subscriber a response to private I don't want to cache it. Else if I still want to cache it I would call the setPrivate on postHandle listener not a preHandle event.

In this case it's not only a performance fix but also a fix or change to support making responses private / nocacheable inside preHandle listener.

Reply:
you are right. lets say it like this?

```suggestion
* Fix for Symfony HttpCache: Do not call store if Response object is not longer cacheable after event listeners. If you use the custom TTL system, this is only a performance improvement, because the TTL of the response would still be 0. With a custom listener that changes the response explicitly to not be cached but does not change s-maxage, this bug might have led to caching responses that should not have been cached
```

Reply:
@dbu sounds good 👍